ArjunaJTA/jta/tests/classes/com/hp/mwtests/ts/jta/recovery/XARecoveryModuleUnitTest.java

```diff
@@ -518,28 +521,25 @@ public XAResource[] getXAResources() throws Exception {
             }
         });
 
-
-        // The first two recovery cycles do nothing with the resource (because phase two is getting the exception)
+        // During the periodicWorkFirstPass recover method is called twice
```

Comment:
The logic of this test is no longer testing the same behaviour, in particular the final assert `assertFalse(rolledback);` has been switched, the original test did `assertTrue(rolledback);` which is what we want because on the third recovery cycle there is no exception and it is only then that we detect that it's an orphan (because the resource says it doesn't know about the xid on the 3rd cycle) and recovery rolls it back and hence `assertTrue(rolledback);` will pass. The code you need, in addition to the resource recover method change, is as follows (note that I have included calls to periodicWorkSecondPass otherwise the revovery module thinks it is `BETWEEN_PASSES` and runs different logic (but we want the test to test the standard logic):

```

        // The first two recovery cycles do nothing with the resource (because phase one is getting the exception)
        // When count reaches 6 it sees that the xid has gone and presumes abort so calls rollback and hence
        // assertTrue(rolledback) passes
        xarm.periodicWorkFirstPass(); // recoverCount is now 2 (and the TMENDRSCAN will have thrown an exception)
        xarm.periodicWorkSecondPass(); // complete the cycle
// I didn't debug why the jndi names assert is failing (I think you figured it out though on the first review)
//        assertTrue(xarm.getContactedJndiNames().contains("test"));
        assertFalse(rolledback); // since recover threw an exception there is no rollback yet

        xarm.periodicWorkFirstPass(); // recoverCount count is now 4 (and the TMENDRSCAN will have thrown an exception)
        xarm.periodicWorkSecondPass(); // complete the cycle
//        assertFalse(xarm.getContactedJndiNames().contains("test"));
        assertFalse(rolledback); // since recover threw an exception there is no rollback yet

        xarm.periodicWorkFirstPass(); // recoverCount count is now 6. On the TMENDRSCAN it will have returned 1 xid and none on the TMSENDRSCAN
        xarm.periodicWorkSecondPass(); // complete the cycle
//        assertTrue(xarm.getContactedJndiNames().contains("test"));
        // since there was a successful recovery cycle (ie no exceptions) which returned an empty list of xids
        // it should have rolled back as an orphan
        assertTrue(rolledback);
```

Reply:
Updated, thanks!